kymatio/scattering2d/backend/numpy_backend.py

```diff
@@ -0,0 +1,178 @@
+# Authors: Edouard Oyallon, Sergey Zagoruyko
+
+import numpy as np
+from collections import namedtuple
+
+BACKEND_NAME = 'numpy'
+
+class Pad(object):
+    def __init__(self, pad_size, input_size, pre_pad=False):
+        """
+            Padding which allows to simultaneously pad in a reflection fashion
+            and map to complex.
+            Parameters
+            ----------
+            pad_size : list of 4 integers
+                size of padding to apply.
+            input_size : list of 2 integers
+                size of the original signal
+            pre_pad : boolean
+                if set to true, then there is no padding, one simply adds the imaginarty part.
+        """
+        self.pre_pad = pre_pad
+        self.pad_size = pad_size
+
+    def __call__(self, x):
+        if self.pre_pad:
+            return x
+        else:
+            np_pad = ((self.pad_size[0], self.pad_size[1]), (self.pad_size[2], self.pad_size[3]))
+            batch_shape = x.shape[:-2]
+            signal_shape = x.shape[-2:]
+            x = x.reshape((-1, 1) + signal_shape)
+            output = np.pad(x, ((0,0), (0,0), np_pad[0], np_pad[1]), mode='reflect')
+            output = output.reshape(batch_shape + output.shape[-2:])
```

Comment:
Same, but on torch backend its -3 instead. Maybe we should wait for the release of complex tensors?

Reply:
I don't understand. Wait to do what?

Reply:
wait until complex tensors are released in Pytorch to refactor this line out to core. 

If this was the torch backend, then we would want `output.shape[-2:]` to be a `output.shape[-3:]`.

Reply:
Like I said in another comment, I don't think all of this reshaping needs to be done in the backend. We should simply be able to specify that the backend deals with arrays of shape `(B, N, M)` *always* (for the torch backends, that would be arrays of shape `(B, N, M, 2)`), and it's up to the frontend wrapper to deal with any other batch shapes.

Reply:
This has now been removed (all batch reshaping is done in frontend).